### PR TITLE
RUN-1800 - collect runai logs

### PR DESCRIPTION
**Pull request by moshe-kabala**

### CODE REVIEW

File: cmd/log/log.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	return intoTar(collectedPaths, fmt.Sprint("runai-logs-", time.Now().Unix()))
+}
+
+func getPodLogsInto(client *client.Client, pod corev1.Pod, containerName string, w io.Writer) error {
+
+	lb := int64(1024 * 1024 * 10) // 10mb
+	podLogOpts := corev1.PodLogOptions{
+		LimitBytes: &lb,
+		Container:  containerName,
+		Previous:   pod.Status.Phase != corev1.PodRunning,
```
Comment (itayvallach) [inline cmd/log/log.go:102]: Need to talk about this, I think the more correct will be:
`pod.Status.Phase == corev1.PodPending`
Meaning the only case we would want the previous container is if the current pod is Pending.
If it's Failed for example we would really want this container and not the previous one. the same for Succeeded and Running, so I guess it leaves only Pending which we would want the previous.

File: cmd/log/log.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/run-ai/runai-cli/pkg/client"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func Command() *cobra.Command {
+
+	var command = &cobra.Command{
+		Use:   "log",
+		Short: "Collect runai logs into tar file",
```
Comment (itayvallach) [inline cmd/log/log.go:23]: Generates a file containing logs from runai system pods

File: cmd/log/log.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+	"github.com/run-ai/runai-cli/pkg/client"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func Command() *cobra.Command {
+
+	var command = &cobra.Command{
+		Use:   "log",
```
Comment (itayvallach) [inline cmd/log/log.go:22]: I wrote you in slack, the command should be (after consulting with Yaron and OmerB):
`generate-log-file`

Need to consider changing the file name as well

File: cmd/log/log.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	if len(collectedPaths) == 0 {
+		return fmt.Errorf("Failed to collect logs")
+	}
+
+	log.Info("End to collect the logs start to create the output file")
+
+	return intoTar(collectedPaths, fmt.Sprint("runai-logs-", time.Now().Unix()))
+}
+
+func getPodLogsInto(client *client.Client, pod corev1.Pod, containerName string, w io.Writer) error {
```
Comment (itayvallach) [inline cmd/log/log.go:96]: can we rename the function? it is not really clear, get pod logs into what/where?

File: cmd/log/log.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+			if client == nil {
+				log.Error("error in getting access to K8S")
+				os.Exit(1)
+			}
+
+			if err := collectLogs(client); err != nil {
+				log.Error(err)
+				os.Exit(1)
+			} else {
+				log.Info("Successfully collect runai logs")
```
Comment (itayvallach) [inline cmd/log/log.go:36]: collected*

File: cmd/log/log.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+func Command() *cobra.Command {
+
+	var command = &cobra.Command{
+		Use:   "log",
+		Short: "Collect runai logs into tar file",
+		Run: func(cmd *cobra.Command, args []string) {
+
+			client := client.GetClient()
+			if client == nil {
+				log.Error("error in getting access to K8S")
```
Comment (itayvallach) [inline cmd/log/log.go:28]: Error accessing Kubernetes cluster

File: cmd/log/log.go (hunk 3 above)
Comment (moshe-kabala) [reply cmd/log/log.go:22]: done

File: cmd/log/log.go (hunk 2 above)
Comment (moshe-kabala) [reply cmd/log/log.go:23]: done

File: cmd/log/log.go (hunk 5 above)
Comment (moshe-kabala) [reply cmd/log/log.go:36]: done

File: cmd/log/log.go (hunk 4 above)
Comment (moshe-kabala) [reply cmd/log/log.go:96]: getPodLogsIntoWriter

File: cmd/generate-log-file/generate-log-file.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		} else {
+			collectedPaths = append(collectedPaths, path)
+			log.Info(i+1, " of ", len(paths), " are done")
+		}
+	}
+	if len(collectedPaths) == 0 {
+		return fmt.Errorf("Failed to collect logs")
+	}
+
+	log.Info("End to collect the logs start to create the output file")
```
Comment (itayvallach) [inline cmd/generate-log-file/generate-log-file.go:92]: Finished collecting logs. Creating the output file...

File: cmd/generate-log-file/generate-log-file.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+			path := "/tmp/log-" + pod.Name
+			if isMultiContainers {
+				path += "--" + container.Name
+			}
+			paths = append(paths, path)
+			podc := pod.DeepCopy()
+			containerName := container.Name
+			go (func() {
+
+				log.Info("Collect logs for pod: ", podc.Name, " container: ", containerName)
```
Comment (itayvallach) [inline cmd/generate-log-file/generate-log-file.go]: Collecting*

File: cmd/generate-log-file/generate-log-file.go (hunk 7 above)
Comment (moshe-kabala) [reply cmd/generate-log-file/generate-log-file.go:92]: done

File: cmd/generate-log-file/generate-log-file.go (hunk 8 above)
Comment (moshe-kabala) [reply cmd/generate-log-file/generate-log-file.go]: done

File: cmd/log/log.go (hunk 1 above)
Comment (moshe-kabala) [reply cmd/log/log.go]: done

**Review by itayvallach — APPROVED** [review]

**Pull request merged**